sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -170,65 +171,75 @@ class EquivalentExpressions {
       // can cause error like NPE.
       (expr.isInstanceOf[PlanExpression[_]] && TaskContext.get != null)
 
-    if (!skip && !addFunc(expr)) {
-      childrenToRecurse(expr).foreach(addExprTree(_, addFunc))
-      commonChildrenToRecurse(expr).filter(_.nonEmpty).foreach(addCommonExprs(_, addFunc))
+    if (!skip && !addExprToMap(expr, map)) {
+      val height = childrenToRecurse(expr).map(addExprTree0(_, map))
+        .reduceOption(_ max _).map(_ + 1).getOrElse(0)
+      map(ExpressionEquals(expr)).height = height
```

Comment:
(My comment is the same with [the @viirya one](https://github.com/apache/spark/pull/33142#issuecomment-871634042)) we can always judge if an expr is a parent of another expr or not from this `height`? It seems this height depends on a `map` state, so a true height value can change after the assignment? For this purpose, we cannot simply use the height of an expression instead?

Reply:
Given that @viirya will remove the sort, I think this issue doesn't matter now (not worse than before). In `addCommonExprs`, I only use this height to do filtering, which should be OK.

Reply:
> For this purpose, we cannot simply use the height of an expression instead?

makes sense. A true height (from root to the furthest leaf) is good enough to quickly check child-parent relationship.